proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,314 @@
+
+  * It could be made not-special, at the cost of more complex machinery. See "Alternatives"
+
+* String interpolation makes things like SQL injection easier
+
+  * I would argue that someone unaware of injection attacks (and libraries designed to prevent these attacks) would be just as likely to manually concatenate a string as interpolate
+
+Alternatives
+------------
+
+* Status quo (discussed in the "Motivation" section)
+
+* Generalize to any type class outputting ``String`` (not just ``Interpolate``)
+
+  * Change the syntax to enable specifying any module defining an ``Interpolate``-compatible interface, similar to ``QualifiedDo`` e.g.
+    ::
+
+      module MyInterpolate where
+
+      class MyInterpolate a where
+        interpolatePrec :: Int -> a -> ShowS
+
+      module Main where
+
+      -- desugars to `("Example " ++) . interpolatePrec 0 foo $ ""`
+      original :: String
+      original = [s|Example ${foo}|]
+
+      -- desugars to `("Example " ++) . MyInterpolate.interpolatePrec 0 foo $ ""`
+      custom :: String
+      custom = [s.MyInterpolate|Example ${foo}|]
+
+* Generalize to any output type
+
+  * Break out a type class representing "Type that an interpolation can result in" and enable specifying any module defining a compatible interface, similar to ``QualifiedDo`` (would also enable swapping out the ``Interpolate`` type class, like the previous bulletpoint). e.g.
+    ::
+
+      {----- built-in to GHC -----}
+
+      class Interpolate s where
+        interpolateRaw :: String -> s -> s
+      class Interpolate s => InterpolateValue s a where
+        interpolatePrec :: Int -> a -> s -> s
+
+      instance Interpolate String where
+        interpolateRaw = (++)
+
+      -- the usual instances
+      instance InterpolateValue String String
+      instance InterpolateValue String Int
+
+      {----- user-defined -----}
+
+      module MySqlQuery where
+
+      data SqlQuery = SqlQuery { queryText :: String, queryVals :: [SqlVal] }
+      instance Interpolate SqlQuery where
+        interpolateRaw q1 (SqlQuery q2 vs) = SqlQuery (q1 <> q2) vs
+      instance InterpolateValue SqlQuery String where
+        interpolatePrec _ s (SqlQuery text vs) = SqlQuery text (SqlString s : vs)
+      instance InterpolateValue SqlQuery Int where
+        interpolatePrec _ x (SqlQuery text vs) = SqlQuery text (SqlInt x : vs)
+
+      module Main where
+
+      -- desugars to `interpolateRaw "SELECT * FROM foo WHERE id = " . interpolatePrec 0 id $ ""`
+      pwnable :: Int -> String
+      pwnable id = [s|SELECT * FROM foo WHERE id = ${id}|]
+
+      -- desugars to `MySqlQuery.interpolateRaw "SELECT * FROM foo WHERE id = " . MySqlQuery.interpolatePrec 0 id $ ""`
+      safe :: Int -> SqlQuery
+      safe id = [s.MySqlQuery|SELECT * FROM foo WHERE id = ${id}|]
```

Comment:
It seems dangerous to ask the programmer to remember to look for and add an additional qualifier to get safe SQL query building. Note that the proposed interaction with `-XOverloadedStrings` would call `fromString` on an interpolation result, and the reason it's safe to give SQL Query types an `IsString` instance is that people generally only use them on literals or things that look like literals.

But if we had interpolations and someone went off to try and do query building via interpolation, GHC would silently let them. This seems like a bad outcome.

Reply:
This is in the Alternatives section, so not part of the main proposal. And this syntax isn't meant to interact with `OverloadedStrings`; I mention this in a sub-bulletpoint.

> the reason it's safe to give SQL Query types an IsString instance is that people generally only use them on literals or things that look like literals

Sure, this is a general remark on string interpolation, not this specific code snippet. You're right, assuming people aren't just using `fromString` around their codebase, `IsString` is currently safe from unsafe building in a way that wouldn't be true with string interpolation.

This specific code snippet actually protects against that, then, by defining a new `Interpolate` class instead of reusing `OverloadedStrings` (again, this `[s.MySqlQuery|` example does _not_ use `OverloadedStrings`).

Reply:
So in this alternative, if a custom interpolator is used and `-XOverloadedStrings` are enabled, `fromString` is _not_ called on the result? That's one thing, but you still have the problem where if the programmer forgets to use the custom interpolator and `-XOverloadedStrings` is on, you will get interpolation into an SQL string.

Reply:
This whole example is incompatible with OverloadedStrings. So if we implemented this alternative ("generalizing the output type"), string interpolation would not use IsString at all

Reply:
Does that then mean you have to select the correct custom interpolator when you switch between `String`, strict `Text`, and lazy `Text`?

Reply:
Again, this is all an Alternative proposal, so I haven't thought in depth about this. But it could select whatever `Interpolate` instance is in scope, with `[s.MySqlQuery|` being analogous to a qualified import